File: Array/Add One To Number.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func main() {

	A := []int{0, 0, 4, 4, 6, 0, 9, 6, 5, 1}
	result := plusOne1(A)
	fmt.Println(result)
}

func plusOne(A []int) []int {
	tempNum := ""
	for i := 0; i < len(A); i++ {
		tempNum += strconv.Itoa(A[i])
	}

	num, _ := strconv.Atoi(tempNum)
	num += 1

	tempNum = strconv.Itoa(num)
	str := strings.Split(tempNum, "")

	var ans []int

	for _, val := range str {
		intNum, _ := strconv.Atoi(val)
		ans = append(ans, intNum)
	}

	return ans
}

func plusOne1(A []int) []int {
	var temp []int
	var ans []int

	carry := 1

	for i := len(A) - 1; i >= 0; i-- {
		sum := A[i] + carry
		carry = sum / 10
		temp = append(temp, sum%10)
	}
	if carry > 0 {
		temp = append(temp, carry)
	}

	idx := len(temp) - 1
	for i := len(A); i >= 0; i-- {
		if temp[idx] == 0 {
			idx--
		}
	}

	for i := idx; i >= 0; i-- {
		ans = append(ans, temp[i])
	}

	return ans
}

File: Pythons Solutions/Add One To Number.py
class Solution:
    def addOneToNum(self, arr):
        temp, carry, ans = [], 1, []

        for i in range(len(arr)-1, -1, -1):
            sum = arr[i] + carry
            carry = sum // 10
            temp.append(sum%10)
            sum = 0
        
        if carry > 0:
            temp.append(carry)

        idx = len(temp)-1
        for i in range(len(arr)-1, -1, -1):
            if temp[i] == 0:
                idx -=1
        
        for i in range(idx, -1, -1):
            ans.append(temp[i])

        return ans

obj = Solution()
arr = [0, 1, 2, 3]
ans = obj.addOneToNum(arr)
print(ans)